corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py

```diff
@@ -0,0 +1,175 @@
+                    print "Progress: {} of {} ({})".format(i, total, round(i / float(total), 2))
+                    print "We can fix {} of {} ({})".format(
+                        num_fixed, i, round(num_fixed / float(i), 2)
+                    )
+                if xform_instance.app_id in new_xmlnss:
+                    num_fixed += 1
+                    # We've already generated a new xmlns for this app
+                    set_xmlns_on_submission(xform_instance, new_xmlnss[xform_instance.app_id], dry_run)
+                    f.write("Set new xmlns on submission {}\n".format(xform_instance._id))
+                else:
+                    app = Application.get(xform_instance.app_id)
+                    forms_without_xmlns = get_forms_without_xmlns(app)
+                    if len(forms_without_xmlns) == 1:
+                        form = forms_without_xmlns[0]
+                        if not xforms_with_real_xmlns_possibly_exist(app._id, form):
+                            num_fixed += 1
+                            new_xmlns = generate_random_xmlns()
+                            new_xmlnss[xform_instance.app_id] = new_xmlns
+                            f.write("New xmlns for form {form.unique_id} in app {app._id} is {new_xmlns}\n".format(
+                                form=form,
+                                app=app,
+                                new_xmlns=new_xmlns
+                            ))
+                            set_xmlns_on_form(form, new_xmlns, app, dry_run)
+                            set_xmlns_on_submission(xform_instance, new_xmlnss[xform_instance.app_id], dry_run)
+                            f.write("Set new xmlns on submission {}\n".format(xform_instance._id))
+
+
+def get_submissions_without_xmlns():
+    # TODO: This view won't get archived forms etc
+    submissions = XFormInstance.get_db().view(
+        'couchforms/by_xmlns',
+        key="undefined",
+        include_docs=False,
+        reduce=False,
+    ).all()
+    total_submissions = len(submissions)
+    submission_id_generator = (s['id'] for s in submissions)
+    doc_generator = (XFormInstance.wrap(i) for i in iter_docs(XFormInstance.get_db(), submission_id_generator))
+    return total_submissions, doc_generator
+
+
+def xforms_with_real_xmlns_possibly_exist(app_id, form):
+    """
+    Return True if there exist xforms submitted to the given app against a form
+    with the same name as the given name that have a real xmlns.
+
+    If this is True, it could indicate that the form had an xlmns, had submissions,
+    then lost the xmlns. If that's the case, we would want to give the submissions
+    this xmlns, not a new random one.
+    """
+    for form_name in form.name.values():
+        query = (FormES()
+                 .term('form.@name', form_name)
+                 .app(app_id)
+                 .filter(NOT(xmlns('undefined'))))
+        if query.count():
+            return True
+    return False
```

Comment:
You could probably memoize this function, as I imagine the affected forms will end up calling this with the same arguments a bunch of times.